src/libsystemd-network/ndisc-router.c

```diff
@@ -821,6 +824,122 @@ int sd_ndisc_router_captive_portal_get_uri(sd_ndisc_router *rt, const char **ret
         return 0;
 }
 
+static int ndisc_get_dns_name(const uint8_t *optval, size_t optlen, char **ret) {
+        _cleanup_free_ char *name = NULL;
+        int r;
+
+        assert(optval || optlen == 0);
+        assert(ret);
+
+        r = dns_name_from_wire_format(&optval, &optlen, &name);
+        if (r < 0)
+                return r;
+        if (r == 0 || optlen != 0)
+                return -EBADMSG;
+
+        *ret = TAKE_PTR(name);
+        return r;
+}
+
+int sd_ndisc_router_encrypted_dns_get_dnr(sd_ndisc_router *rt, ResolverData **ret) {
+        uint8_t *nd_opt_encrypted_dns = NULL;
+        size_t length;
+        int r;
+
+        assert_return(rt, -EINVAL);
+        assert_return(ret, -EINVAL);
+
+        r = sd_ndisc_router_option_is_type(rt, SD_NDISC_OPTION_ENCRYPTED_DNS);
+        if (r < 0)
+                return r;
+        if (r == 0)
+                return -EMEDIUMTYPE;
+
+        r = sd_ndisc_router_option_get_raw(rt, (void *)&nd_opt_encrypted_dns, &length);
+        if (r < 0)
+                return r;
+
+        assert(length % 8 == 0);
+        if (length == 0)
+                return -EBADMSG;
+
+        uint8_t *optval = &nd_opt_encrypted_dns[2];
+        size_t optlen = 8 * length - 2; /* units of octets */
+        size_t offset = 0;
+
+        _cleanup_(dnr_resolver_data_free_allp) ResolverData *res = NULL;
+        res->priority = be16toh(optval[offset]);
```

Comment:
Looks like networkd can segfault here on receiving weird packets with
```sh
==348==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x00000097c08f bp 0x7fffe8699c30 sp 0x7fffe8699920 T0)                   [226/1861]
==348==The signal is caused by a WRITE memory access.                                                                                                         
==348==Hint: address points to the zero page.                                                                                                                 
    #0 0x97c08f in sd_ndisc_router_encrypted_dns_get_dnr /home/vagrant/systemd-asan/build/../src/libsystemd-network/ndisc-router.c:871:23                     
    #1 0x85e275 in ndisc_router_process_encrypted_dns /home/vagrant/systemd-asan/build/../src/network/networkd-ndisc.c:1071:13                                
    #2 0x8485e0 in ndisc_router_process_options /home/vagrant/systemd-asan/build/../src/network/networkd-ndisc.c:1152:29                                      
    #3 0x83f368 in ndisc_router_handler /home/vagrant/systemd-asan/build/../src/network/networkd-ndisc.c:1415:13                                              
    #4 0x83d470 in ndisc_handler /home/vagrant/systemd-asan/build/../src/network/networkd-ndisc.c:1445:21                                                     
    #5 0xa964c1 in ndisc_callback /home/vagrant/systemd-asan/build/../src/libsystemd-network/sd-ndisc.c:42:9                                                  
    #6 0xa95822 in ndisc_handle_datagram /home/vagrant/systemd-asan/build/../src/libsystemd-network/sd-ndisc.c:197:9                                          
    #7 0xa924e9 in ndisc_recv /home/vagrant/systemd-asan/build/../src/libsystemd-network/sd-ndisc.c:250:16                                                    
    #8 0x7f2fe6ea339c in source_dispatch /home/vagrant/systemd-asan/build/../src/libsystemd/sd-event/sd-event.c:4187:21                                       
    #9 0x7f2fe6ea13c4 in sd_event_dispatch /home/vagrant/systemd-asan/build/../src/libsystemd/sd-event/sd-event.c:4808:21                                     
    #10 0x7f2fe6ea6ebd in sd_event_run /home/vagrant/systemd-asan/build/../src/libsystemd/sd-event/sd-event.c:4869:21                                         
systemd-networkd.service: Main process exited, code=exited, status=1/FAILURE                                                                                  
    #11 0x7f2fe6ea7e49 in sd_event_loop /home/vagrant/systemd-asan/build/../src/libsystemd/sd-event/sd-event.c:4891:21                                        
    #12 0x417339 in run /home/vagrant/systemd-asan/build/../src/network/networkd.c:112:13                                                                     
    #13 0x41400d in main /home/vagrant/systemd-asan/build/../src/network/networkd.c:119:1                                                                     
    #14 0x7f2fe5449b89 in __libc_start_call_main (/lib64/libc.so.6+0x27b89) (BuildId: f888be5f5e7d58e04cabb8c675c7ab94e77dd68c)                               
    #15 0x7f2fe5449c4a in __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x27c4a) (BuildId: f888be5f5e7d58e04cabb8c675c7ab94e77dd68c)                        
    #16 0x412a24 in _start (/usr/lib/systemd/systemd-networkd+0x412a24) (BuildId: 1427615bd07bef37e99485a00c0bb50aa4465edf)                                   
AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV /home/vagrant/systemd-asan/build/../src/libsystemd-network/ndisc-router.c:871:23 in sd_ndisc_router_encrypted_dns_get_dnr
==348==ABORTING
systemd-networkd.service: Main process exited, code=exited, status=1/FAILURE
systemd-networkd.service: Failed with result 'exit-code'.
systemd-networkd.service: Scheduled restart job, restart counter is at 10.
systemd-networkd.service: Start request repeated too quickly.
systemd-networkd.service: Failed with result 'exit-code'.
Failed to start systemd-networkd.service.
systemd-networkd.socket: Failed with result 'service-start-limit-hit'.
```

Reply:
"can" lol. Yeah that should definitely be a malloc and not "= NULL". Yiiikes.